include/ringmesh/basic/attributes.h

```diff
@@ -0,0 +1,1024 @@
+        index_t element_size_; // size of element in byte
+    };
+
+    template < class T >
+    class TypedStore : public Store
+    {
+    public:
+        TypedStore() : Store( index_t( sizeof( T ) ) )
+        {
+        }
+
+        bool elements_type_matches( const std::string& type_name ) const final
+        {
+            return type_name == element_typeid_name();
+        }
+
+        std::string element_typeid_name() const override
+        {
+            return typeid( T ).name();
+        }
+    };
+
+    template < class T >
+    class VectorStoreBase : public TypedStore< T >
+    {
+    public:
+        VectorStoreBase() = default;
+
+        void resize( index_t new_size ) final
+        {
+            this->size_ = new_size;
+            vector_.resize( new_size );
+        }
+
+        const void* data() const final
+        {
+            return vector_.data();
+        }
+
+        void* data() final
+        {
+            return vector_.data();
+        }
+
+        const void* element( index_t element ) const final
+        {
+            return &this->vector_[element];
+        }
+        void* element( index_t element ) final
+        {
+            return &this->vector_[element];
+        }
+        void clear() final
+        {
+            vector_.clear();
+        }
+
+        void compress( const std::vector< index_t >& old2new ) final
+        {
+            ringmesh_assert( old2new.size() < this->size() );
+            for( index_t new_id = 0; new_id < old2new.size(); new_id++ )
+            {
+                index_t old_id = old2new[new_id];
+                if( old_id == NO_ID || old_id == new_id )
+                {
+                    continue;
+                }
+                ringmesh_assert( old_id < this->size() );
+                vector_[new_id] = vector_[old_id];
+            }
+        }
+        /**
+        * \brief Applies a permutation to the stored attributes.
+        * \details Applying a permutation to the data is equivalent
+        *  to:
+        * \code
+        * for(i=0; i<permutation.size(); i++) {
+        *    data2[i] = data[permutation[i]]
+        * }
+        * data = data2 ;
+        * \endcode
+        * But it is done in-place.
+        * \param[in] permutation the permutation.
+        *  It is temporarily changed during execution of the
+        *  function, but identical to the input on exit.
+        * \note This function uses memcpy(). If required, it
+        *  can be overloaded in derived classes.
+        */
+        void apply_permutation(
+            const std::vector< index_t >& permutation ) final
+        {
+            std::vector< T > tmp_vector;
+            tmp_vector.resize( vector_.size() );
+            for( auto p : range( permutation.size() ) )
+            {
+                tmp_vector[p] = vector_[permutation[p]];
+            }
+            vector_ = tmp_vector;
+        }
+
+    protected:
+        std::vector< T > vector_;

```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Make this member variable "private". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=cpp%3AS3656)